check_stability.py

```diff
@@ -483,31 +482,69 @@ class LogHandler(reader.LogHandler):
         Subclasses reader.LogHandler.
         """
         def __init__(self):
-            self.results = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
+            self.results = []
+
+        def find_or_create_test(self, data):
+            for test in self.results:
+                if test["name"] == data["test"]:
+                    return test
+
+            test = {
+                "name": data["test"],
+                "subtests": [],
+                "status": {"OK": 0, "FAIL": 0}
+            }
+            self.results.append(test)
+            return test
+
+        def find_or_create_subtest(self, data):
+            test = self.find_or_create_test(data)
+
+            for subtest in test["subtests"]:
+                if subtest["name"] == data["subtest"]:
+                    return subtest
+
+            subtest = {
+                "name": data["subtest"],
+                "status": {"PASS": 0, "FAIL": 0},
+                "messages": set()
+            }
+            test["subtests"].append(subtest)
+
+            return subtest
 
         def test_status(self, data):
-            self.results[data["test"]][data.get("subtest")][data["status"]] += 1
+            subtest = self.find_or_create_subtest(data)
+            subtest["status"][data["status"]] += 1
+            if (data.get("message")):
+                subtest["messages"].add(data["message"])
 
         def test_end(self, data):
-            self.results[data["test"]][None][data["status"]] += 1
+            test = self.find_or_create_test(data)
+            test["status"][data["status"]] += 1
 
 
 def is_inconsistent(results_dict, iterations):
     """Return whether or not a single test is inconsistent."""
-    return len(results_dict) > 1 or sum(results_dict.values()) != iterations
+    values = results_dict.values()
+    nonzero_results = filter(lambda count: count > 0, values)
```

Comment:
I think this will disappear when we use a `defaultdict`, but in general using comprehensions is preferred over using a filter in cases like this e.g. `any(item for item in values if item > 0)`.